File: cs/1-two-sum.cs
using System;
class TwoSum {
        static void Main(string[] args) {
                // int[] nums = new int[5] { 2, 5, 46, 14, 9 };
                int[] nums = new int[2] { 3, 3 };
                int target = 11;

                int[] result = TwoSumTwoPassHashtable(nums, target);
                Array.ForEach(result, Console.WriteLine);
        }

        // O(n^2) time - O(1) space
        public static int[] TwoSumBruteForce(int[] nums, int target) {
                int size = nums.Length;
                int[] indices = new int[2];

                for (int i = 0; i < size - 1; i++) {
                        for (int j = i + 1; j < size; j++) {
                                if (nums[i] + nums[j] == target) {
                                        indices[0] = i;
                                        indices[1] = j;

                                        return indices;
                                }
                        }
                }

                throw new ArgumentException("Not found");
        }

        // O(n) time & space
        public static int[] TwoSumTwoPassHashtable(int[] nums, int target) {
                int size = nums.Length;
                int[] indices;
                var nums_hash = new System.Collections.Hashtable(); // num & index
                // var nums_hash = new System.Collections.Generic.Dictionary<int, int>();

                // generate hash table from array
                for (int i = 0; i < size; i++) {
                        int num = nums[i];

                        if (!nums_hash.ContainsKey(num)) {
                                nums_hash.Add(num, i);
                                continue;
                        }

                        // corner case: duplicated key
                        if (num == target / 2) {
                                indices = new int[2] { (int)nums_hash[num], i };
                                return indices;
                        }
                }

                // find the complement
                foreach (int num in nums_hash.Keys) {
                        int complement = target - num;
                        if (nums_hash.ContainsKey(complement) && nums_hash[complement] != nums_hash[num]) {
                                indices = new int[2] { (int)nums_hash[num], (int)nums_hash[complement] };
                                return indices;
                        }
                }

                throw new ArgumentException("Not found");
        }

        // O(n) time & space
        public static int[] TwoSumTwoPassDictionary(int[] nums, int target) {
                int size = nums.Length;
                int[] indices;
                var nums_hash = new System.Collections.Generic.Dictionary<int, int>(); // num & index

                // generate dictionary from array
                for (int i = 0; i < size; i++) {
                        int num = nums[i];

                        if (!nums_hash.ContainsKey(num)) {
                                nums_hash.Add(num, i);
                                continue;
                        }

                        // corner case: duplicated key
                        if (num == target / 2) {
                                indices = new int[2] { nums_hash[num], i };
                                return indices;
                        }
                }

                // find the complement
                foreach (int num in nums_hash.Keys) {
                        int complement = target - num;
                        if (nums_hash.ContainsKey(complement) && nums_hash[complement] != nums_hash[num]) {
                                indices = new int[2] { nums_hash[num], nums_hash[complement] };
                                return indices;
                        }
                }

                throw new ArgumentException("Not found");
        }
}

File: cpp/1-two-sum.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

// HELPER
void printHashMap(unordered_map<int, int> hashmap)
{
	for (const auto &n : hashmap)
	{
		std::cout << "Key:[" << n.first << "] Value:[" << n.second << "]\n";
	}
}

void printVector(vector<int> const &a)
{
	cout << "The vector elements are : ";

	for (int i = 0; i < a.size(); i++)
		cout << a.at(i) << ' ';
}

// 0(n^2) time - O(1) space
vector<int> twoSumBruteForce(vector<int> &nums, int target)
{
	const int size = nums.size();
	vector<int> indices;

	for (int i = 0; i < size - 1; i++)
	{
		for (int j = j + 1; j < size; j++)
		{
			if (nums[i] + nums[j] == target)
			{
				// indices.push_back(i);
				// indices.push_back(j);
				indices = {i, j};
			}
		}
	}

	return indices;
}

// O(n) time - O(n) space
vector<int> twoSumOnePassHashtable(vector<int> nums, int target)
{
	vector<int> indices;
	unordered_map<int, int> numsHashtable; //num & index

	for (int i = 0; i < nums.size(); ++i)
	{
		int complement = target - nums[i];

		if (numsHashtable.find(complement) != numsHashtable.end())
		{
			indices = {numsHashtable[complement], i};
			return indices;
		}

		// numsHashtable.insert({nums[i], i});
		numsHashtable[nums[i]] = i;
	}

	return indices;
}

int main()
{
	vector<int> nums{3, 3, 4};
	int target = 61;

	vector<int> result = twoSumOnePassHashtable(nums, target);
	printVector(result);
}